linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/yarn/YarnResourceRequester.java

```diff
@@ -0,0 +1,426 @@
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.linkis.manager.common.errorcode.ManagerCommonErrorCodeSummary.*;
+
+public class YarnResourceRequester implements ExternalResourceRequester {
+  private static final Logger logger = LoggerFactory.getLogger(YarnResourceRequester.class);
+
+  private final String HASTATE_ACTIVE = "active";
+
+  private ExternalResourceProvider provider = null;
+  private final Map<String, String> rmAddressMap = new ConcurrentHashMap<>();
+
+  private String getAuthorizationStr() {
+    String user = (String) provider.getConfigMap().getOrDefault("user", "");
+    String pwd = (String) provider.getConfigMap().getOrDefault("pwd", "");
+    String authKey = user + ":" + pwd;
+    return Base64.getMimeEncoder().encodeToString(authKey.getBytes());
+  }
+
+  @Override
+  public NodeResource requestResourceInfo(
+      ExternalResourceIdentifier identifier, ExternalResourceProvider provider) {
+    String rmWebHaAddress = (String) provider.getConfigMap().get("rmWebAddress");
+    this.provider = provider;
+    String rmWebAddress = getAndUpdateActiveRmWebAddress(rmWebHaAddress);
+    logger.info("rmWebAddress: " + rmWebAddress);
+    String queueName = ((YarnResourceIdentifier) identifier).getQueueName();
+
+    String realQueueName = "root." + queueName;
+
+    return LinkisUtils.tryCatch(
+        () -> {
+          Pair<YarnResource, YarnResource> yarnResource =
+              getResources(rmWebAddress, realQueueName, queueName);
+          getResources(rmWebAddress, realQueueName, queueName);
+
+          CommonNodeResource nodeResource = new CommonNodeResource();
+          nodeResource.setMaxResource(yarnResource.getKey());
+          nodeResource.setUsedResource(yarnResource.getValue());
+          return nodeResource;
+        },
+        t -> {
+          throw new RuntimeException(t);
+        });
+  }
+
+  public Optional<YarnResource> maxEffectiveHandle(
+      Optional<JsonNode> queueValue, String rmWebAddress, String queueName) {
+    try {
+      JsonNode metrics = getResponseByUrl("metrics", rmWebAddress);
+      long totalMemory = metrics.path("clusterMetrics").path("totalMB").asLong();
+      long totalCores = metrics.path("clusterMetrics").path("totalVirtualCores").asLong();
```

Comment:
can parse `clusterMetrics ` once and used for `totalMemory ` and `totalCores ` ?

Reply:
Yeah , thanks for your good advice. I have modified it .